### PR TITLE
implement optional tracking event callbacks for external observers

### PR DESCRIPTION
**Pull request by yitao-li**
Signed-off-by: Yitao Li <yitao@rstudio.com>

closes #3501 

## What changes are proposed in this pull request?

Optional callbacks related to the proposal in https://github.com/mlflow/mlflow/issues/3494

## How is this patch tested?

Existing unit tests will suffice

### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: Local serving, model deployment tools, spark UDFs
- [ ] `area/server-infra`: MLflow server, JavaScript dev server
- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, JavaScript, plotting
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [x] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

**Comment by javierluraschi** [comment]
Test failures are probably unrelated, but let me rerun them to see if we get a clean pass for:

```
 AssertionError: assert 'Invalid filter' in "Invalid clause(s) in filter string: 'IN'. Expected search filter with single comparison operator. e.g. name='myModelName'"
```

**Review by javierluraschi — APPROVED** [review]

**Comment by javierluraschi** [comment]
We try to mark R PRs with the label `language/r`, added.

**Comment by javierluraschi** [comment]
@yitao-li can you try rebasing from master? I don't think is possible for this change to break the Python tests, but we want to have tests passing before we merge.

File: mlflow/R/mlflow/R/tracking-observer.R
Comment (smurching) [inline mlflow/R/mlflow/R/tracking-observer.R:6]: Should we catch exceptions thrown by observers/log warnings? It seems like exceptions thrown by observers shouldn't fail user code

File: mlflow/R/mlflow/R/tracking-globals.R
Comment (smurching) [inline mlflow/R/mlflow/R/tracking-globals.R:34]: Should this be named e.g. "set_tracking_uri" for parity with events like "log_metric"?

File: mlflow/R/mlflow/R/tracking-globals.R
Comment (smurching) [inline mlflow/R/mlflow/R/tracking-globals.R:4]: Out of curiosity, why do we register the callback on `set_active_run_id` rather than on `mlflow_start_run`?

**Review by smurching — COMMENTED** [review]
@yitao-li thanks for the PR! Could we add some tests e.g. under a file called test-tracking-observer.R under https://github.com/mlflow/mlflow/tree/c50b04b53d7d7fcd2b78391fbe95a064205697be/mlflow/R/mlflow/tests/testthat for this new functionality, and document it so that it appears in the R API docs?

See https://github.com/mlflow/mlflow/blob/master/CONTRIBUTING.rst#writing-docs for guidelines on writing docs - our R doc generation process is a bit wonky, you might be able to add a section describing this MLflowObservers option and its usage in [this file](https://github.com/mlflow/mlflow/blob/HEAD/mlflow/R/mlflow/document.R#L42)

I also just want to run this overall proposal by other committers - it seems reasonable & useful to have such callbacks, but just want to make sure there aren't any cons I'm missing. The only downsides I can immediately think of are that it may make it more costly (from an engineering time perspective) to add new fluent APIs if we need to call a corresponding callback.

The interface of passing callback data as a list is also interesting - I'm not an R expert so I'm not sure how user-friendly it is to receive callback information as a list :) but one definite advantage seems to be that we can add new arguments to fluent APIs without breaking existing callbacks (because the new argument will just be a new key-value pair in the dict). 

If you could write up (e.g. in google docs or a github gist) a brief writeup describing the new interfaces in this PR & their pros/cons, along with alternatives considered, it'd also help expedite the review process - thanks!

**Comment by yitao-li** [comment]
@smurching Thanks for the feedback. I'll be more than happy to create R documentation for the callbacks.

Question about documentation: I see the content of R-api.rst is based on what's in mlflow/R/mlflow/Reference_Manual_mlflow.md, but Reference_Manual_mlflow.md is in .gitignore (!!) but there is no Rmd file in the repo that can generate Reference_Manual_mlflow.md.
Can we please update the exact steps needed for creating R documentation (other than "make rdocs")?

**Comment by yitao-li** [comment]
@smurching The writeup for this is in https://gist.github.com/yitao-li/376bf4e4a1ad62f92311679e9567bd1d
but as previously mentioned I would like some clarification on how exactly should documentation be created for mlflow the R package -- currently it's quite unclear

**Comment by yitao-li** [comment]
Tests are passing now.

File: mlflow/R/mlflow/R/tracking-globals.R
Comment (javierluraschi) [reply mlflow/R/mlflow/R/tracking-globals.R:34]: My personal preference is to prefix absolutely everything, we've done that in packages like `sparklyr` (as in `spark_`, `ml_`, etc), `tensorflow` (as in `tf_`, etc), and more recently `torch` (as in `torch_`, etc.). The reasoning, not having a prefix in R packages is expensive... Once you define something outside of a package, you need to figure out how that might potentially interact with other packages.

For instance, for `log_metric`, we might have to consider how that interacts with other model logging methods from the R community. So far, I don't think there are any package clashes, but one day this could change. Would be nice for instance if one could `log_metric` without using mlflow, then connect mlflow, and the code won't break.

So net-net, I would encourage we keep prefixes in most functions, unless we feel strongly a function is meant to be top-level and deserves the extra work of potentially interfacing with other packages.

File: mlflow/R/mlflow/R/tracking-observer.R
Comment (javierluraschi) [reply mlflow/R/mlflow/R/tracking-observer.R:36]: I'm 50/50 on this one, since I'm not completely the users of this function should be considered users. Closer description would be developers, which can probably handle just find the exception.

**Pull request merged**
